text/0000-time-improvements.md

```diff
@@ -0,0 +1,350 @@
+2015 would produce September 31, 2015, which is also not a valid date.
+
+Certain human descriptions of durations, like "1 month and 35 days"
+do not make sense, and human descriptions like "1 month and 5 days" have
+ambiguous meaning when used in operations (do you add 1 month first and then
+5 days or vice versa).
+
+
+For these reasons, this RFC does not attempt to define a human duration with
+fields for years, days or months. Such a duration would be difficult to use
+in operations without hard-to-remember ordering rules.
+
+For these reasons, this RFC does not propose APIs related to human concepts
+dates and times. It is intentionally forwards-compatible with such
+extensions.
+
+---
+
+Finally, many APIs that **take** a `Duration` can only do something useful with
+positive values. For example, a timeout API would not know how to wait a
+negative amount of time before timing out. Even discounting the possibility of
+coding mistakes, the problem of system clock time travel means that programmers
+often produce negative durations that they did not expect, and APIs that
+liberally accept negative durations only propagate the error further.
+
+As a result, this RFC makes a number of simplifying assumptions that can be
+relaxed over time with additional types or through further RFCs:
+
+It provides convenience methods for constructing Durations from larger units
+of time (minutes, hours, days), but gives them names like
+`Duration.from_standard_hour`. A standard hour is always 3600 seconds,
+regardless of leap seconds.
+
+It provides APIs that are expected to produce positive `Duration`s, and expects
+that APIs like timeouts will accept positive `Durations` (which is currently
+the case in Rust's standard library). These APIs help the programmer discover
+the possibility of system clock time travel, and either handle the error explicitly,
+or at least avoid propagating the problem into other APIs (by using `unwrap`).
+
+It separates monotonic time (`Instant`) from time derived from the system
+clock (`SystemTime`), which must account for the possibility of time travel.
+This allows methods related to monotonic time to be uncaveated, while working
+with the system clock has more methods that return `Result`s.
+
+This RFC does not attempt to define a type for calendared DateTimes, nor does it
+directly address time zones.
+
+# Proposal
+
+## Types
+
+```rust
+pub struct Instant {
+  secs: u64,
+  nanos: u32
+}
+
+pub struct SystemTime {
+  secs: u64,
+  nanos: u32
+}
+
+pub struct Duration {
+  secs: u64,
+  nanos: u32
+}
+```
+
+### Instant
+
+`Instant` is the simplest of the types representing moments in time. It
+represents an opaque (non-serializable!) timestamp that is guaranteed to
+be monotonic when compared to another `Instant`.
+
+> In this context, monotonic means that a timestamp created later in real-world
+> time will always be larger than a timestamp created earlier in real-world
+> time.
```

Comment:
Consider a 4ghz processor. It completes four cycles per nanosecond; if getting the "now" Instant is single-cycle (as rdtsc can be, for example), then it most certainly can happen.